### DISCUSSION TITLE
brew upgrade <package> just upgrades everything

### DISCUSSION THREAD

**Discussion by frequentredundancy** [Homebrew/discussions#3126] [Everyday usage]
So I wanted to try and upgrade a single package, specifically `php`. At the time of writing, that's PHP 8.1.4
So the command I entered was `brew upgrade php` since the brew documentation/help text lead me to believe this would ONLY update the single package.

However, brew seems to just completely ignore this and decides to upgrade everything anyway? Can someone please explain why? This seems like incredibly bad design, since all I wanted to do was upgrade PHP 8.1 a few patch versions which realistically would take 20-30 seconds max, and instead a number of completely unrelated packages also got upgraded leading to my upgrade of "only php 8.1" taking well over 20+ minutes. Because I'm on an M1 Mac, a lot of these packages have to be built from source which takes a considerable amount of time and absolutely destroys battery life because some of them are absolutely huge and take upwards of 10+ minutes to compile.

I understand dependencies of PHP 8.1 needing to be upgraded since, well they are dependencies that it relies on. But brew seriously just decides it's going to upgrade everything, including packages that I'm confident have very little to do with PHP 8.1. For example;
- Nginx
- Python 2 AND 3
- Httpd
- PHP 7.4 and 8.0
- Other utility packages I have like youtube-dl for example.

I'm not sure if I'm using the command wrong, but I seem to be using it exactly as the documentation suggests and it seems to just not work any different from just writing `brew upgrade`. 

If this is the case, then what exactly is the point in specifying the package in the command in the first place?

**Comment by SMillerDev** [comment]
> Because I'm on an M1 Mac, a lot of these packages have to be built from source which takes a considerable amount of time and absolutely destroys battery life because some of them are absolutely huge and take upwards of 10+ minutes to compile.

There is almost nothing that should require this on M1 unless you installed in a non-default location.

The explanation for your initial question is here though:
https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-or-brew-install-formula-also-upgrade-a-bunch-of-other-stuff

---
TL;DR install in the default location and regularly update your software and you won't have this issue.

**Reply by heyalexchoi** [reply]
truly wild behavior. would expect a warning that xyz upgrades would be triggered, and maybe a tool to help find a version to install that would not require upgrades.

**Reply by chorner** [reply]
I agree this is wacky.  IMO it's the development approach favored by the ruby community: optimize for maintainer ease, ignore the user experience.

**Reply by Eugeny** [reply]
Since there's still lots of confusion going on in other replies, it might be useful to explicitly note that the root cause is that unlike every other package manager, Homebrew needs to _**upgrade reverse dependencies as well**_, including the dependants of the transitive dependencies of whatever you're upgrading

**Reply by SMillerDev** [reply]
It's a side-effect of Homebrew's goal to keep everything on the latest upstream version. There is at least one other package manager that does this, and it even says that partial updates are completely unsupported. While brew just updates whatever it needs to make it work.

[It's Arch btw](https://wiki.archlinux.org/title/System_maintenance#Partial_upgrades_are_unsupported)

**Reply by dash-xd** [reply]
linuxbrew/homebrew is portable but it's admittedly not many people's first choice on linux. my servers are often debian or ubuntu, and now nixos. my main non-mac local dev env is arch. you can use pacman wherever you have arch i.e. wsl arch, arch vm, arch docker; arch and the aur are old and tested, although you should verify aur packages are legit. pacman lets me rollback broken installs with ease. i keep an arch laptop as my main dev laptop that is "set up correctly" for my various dev environments+pipelines. all my macs have been total cluster-Fs. i can see why a certain type of dev might reach for homebrew first to keep things homogenous but those same reasons are why devs are reaching for nix today which is deterministic and doesn't have these "gotchas". i tried snap for a little bit but now only use it very occasionally.

moving forward i will keep an arch VM or arch container that mimics my arch laptop environment.

i can see how all of this is a "skill issue" for not starting with homebrew, but i am admittedly not a mac dev and just love the OS as being my "central brain" for everything i do rather than my central  local dev environment, which is on my arch laptop.

**Reply by Eugeny** [reply]
bro literally went "I use arch btw"

**Reply by dash-xd** [reply]
id like you to read it more like "i use debian, ubuntu, and nixos" in prod (and alpine for tiny things),  so mac dev ci/cd pipeline with homebrew is unbearable. im  switching to nixos locally as soon as its feasible because aur doesnt have the stability i want. if i dont like nix then ill use fedora or centos. arch local is not ideal

**Reply by brandonzylstra** [reply]
It seems like a pretty significant UX shortcoming that Homebrew doesn't warn the user that upgrading `php` will actually upgrade 75 other things.  Does it not have the dependency graph?  Would it be so hard to detect this and ask the user first?

**Reply by p-linnane** [reply]
1. Regularly updating your system will not upgrade 75 other things. If you do not update regularly, that's a digital hygiene failing on the user's end.
2. What do you propose happens in this situation you're describing? You want to upgrade `php`. Are you just going to never upgrade it if too many other packages need to be upgraded as well?

**Reply by quackerex** [reply]
I see huge deps trees as the main issue.

Take php for example, it has about 50 deps while go has none. I don't know if this is apple to apple comparison but why cant php do what go do.

If we don't have huge deps tree == upgrade happens quicker == no complaints

**Reply by SMillerDev** [reply]
> Take php for example, it has about 50 deps while go has none. I don't know if this is apple to apple comparison but why cant php do what go do.

Why doesn't Go have the syntax that PHP does, or the strengths that PHP does? Because they are two different languages designed to do different things.

One of the goals when making the Go language was to always compile binaries without dependencies. One of the goals for PHP is to build on working systems (so depend on other tools).

In addition to that, dependency trees are mostly determined by the maintainers of the tool/language. Homebrew has very little influence over it, so is the wrong place to complain about this. For PHP you should check https://externals.io

**Reply by quackerex** [reply]
> Homebrew has very little influence over it, so is the wrong place to complain about this.

Homebrew does have control over deps and could bottle a minimal variant of php.

There is issue about creating minimal variant of formulas: https://github.com/Homebrew/brew/issues/19789#issuecomment-2820498328. Unfortunately formula variants are unlikely to come to homebrew core.

**Reply by SMillerDev** [reply]
> Homebrew does have control over deps and could bottle a minimal variant of php.

Only optional dependencies. But it also has to be useful to people. We could drop the pcre dependency and let people live without regex, drop curl and stop php from connecting to the internet, but would that be useful to the majority of users? If I look at what my PHP code does, I'd say it would not be.

**Reply by p-linnane** [reply]
I absolutely guarantee the minimal version of PHP you want would be different from the minimal version of PHP others want. We don't have the ability to build a ton of different versions of the same formula in various configurations.

**Comment by aeamaea** [comment]
I have it installed in the "default location", I upgrade it pretty regularly (actually because it just downloads every single forumula every time I make the mistake of installing _any new formula_ ) and tried to install GPG. It's been installing newovim, erlang, and ever other gosh darn thing it can get its hands on. This should NOT be default behaviour, and if it's going to do something like this, it should at least warn the user and give them the option of opting out. Sometimes, we really don't need to burn our battery trying to install what we think might be just one package and a couple dependencies. I fail to see how how neovim and erlang formulas are somehow a dependency for gpg. It's just upgrading/re-installing everthing and that's not a good trip.

**Reply by SMillerDev** [reply]
> actually because it just downloads every single forumula every time I make the mistake of installing any new formula 

It doesn't, as is explained in the link I posted.

> I fail to see how how neovim and erlang formulas are somehow a dependency for gpg.

Checking the dependency tree of all three of those I suspect the following:
- gpg depends on gettext directly and openssl@1.1 indirectly
- both were outdated on your system
- updating them resulted in updating all software that depends on them, which includes neovim through gettext and erlang through openssl@1.1

**Comment by konnano** [comment]
Upgrade with brew install php instead of brew upgrade php

**Reply by ivancuric** [reply]
This is just bad API design

**Comment by melkamar** [comment]
I found a related discussion here https://github.com/orgs/Homebrew/discussions/57 but that was locked, so I will add my 2c here.

I have just run a `brew install jpegoptim`, which resulted in a 15-minute-long download and installation of _a lot of stuff_. It ended up failing on some `node` install step, but I can live with that, I just need to get working with PHP quickly. But, turns out, installing `jpegoptim` caused my PHP to upgrade 8.1 -> 8.3, so now my PHP dev environment does not work.

Based on the linked discussion, this is just how things work with reverse dependency resolution, and I am not trying to change that philosophy. What I really think, though, is that there should be a default behavior of warning the user and requiring a confirmation (in an interactive session, to avoid the "but what about CI argument"). 

Brew must know what it will do with what packages, so it should not be very difficult to display the proposed changes to the user and just ask "are you ok with this"?


(for random people who might stumble here -- the command to revert PHP at least is `brew unlink php && brew link --overwrite --force php@8.1`)

**Reply by dash-xd** [reply]
it's the soydev mentality of feigning ignorance towards sensible CI/CD principles, and also the reason why the brew dev wrote a blogpost about why he got rejected from Google. probably a higher up there who specifically hates this "feature" of brew and the philosophy behind it so they decided to hire from India instead. 

solution: if you're on a silicon mac, just use docker. if you need virtualization use UTM. there is absolutely no reason to use homebrew if you're a poweruser except for dev tools on the host machine. i've made this mistake for the last time

**Reply by Eugeny** [reply]
> if you're on a silicon mac, just use docker

imagine having to run a linux VM instead of a native app, peak soydev mentality

**Reply by dash-xd** [reply]
i just deploy my container to cloud run or fly.io . soydev uses ruby-based package mgr in the first place

**Reply by dash-xd** [reply]
also docker was made by golang-minded devs for golang-minded devs. if you're really going to call golang devs soydevs you live in an alternate reality

**Reply by dash-xd** [reply]
> imagine having to run a linux VM


you have quite the extensive git repo history to make the mistake of thinking a container is a VM

**Reply by Eugeny** [reply]
Docker Desktop on macOS just runs docker in an actual Linux VM.

**Reply by dash-xd** [reply]
that's good to know. personally i use docker-compose and docker from the cli, also planning on switching to podman and alias it to docker as a drop-in very soon.

**Comment by quackerex** [comment]
My solution to this is remove anything I am not using currently.
So If I am not using python@3.11 remove that for now and install whenever I need that again.
And run `brew autoremove` to remove remove any deps that are no longer needed.

This avoids the huge deps trees homebrew needs to upgrade.

**Reply by dash-xd** [reply]
use miniconda on mac and avoid this entirely at least for python

**Comment by dash-xd** [comment]
if you're on a silicon mac just use UTM and/or docker. there is absolutely no reason to use homebrew if you're a poweruser except for small dev tools on the host machine. the silicon chips handle containers and vms like a champ. i've made this mistake for the last time

**Reply by SMillerDev** [reply]
Why waste CPU cycles emulating something when you can run it directly on the hardware?

**Reply by dash-xd** [reply]
because brew is "yet another package manager" i don't want to deal with. it has no place in my pipeline aside from being a convenience tool on mac, and i am realizing it's not convenient. i care about my user / customer experience, not my personal experience in my dev environment. intel might not be able to manage this load though.

here is an unrelated reason why you should use virtualization if you're trying to run a "prod-like" environment on your silicon chips : https://www.phoronix.com/news/Apple-M1-M1RACLES

"For the moment the only workaround/mitigation is running your software within a virtual machine where hypervisors currently disable access by the VM to the s3_5_c15_c10_1 register"

if you're in a situation where you think "i need to have the latest rolling bleeding edge stack on baremetal silicon" consider the vulnerability. i don't think there will be any holding back nix and nixos from dominating in the next few years either

**Reply by p-linnane** [reply]
@dash-xd Posting a 3 and a half year old vulnerability as a reason to never install anything on a Mac is definitely an interesting take.

All you're doing is coming in here to try and start problems and use discriminatory language for some sad reason. You aren't welcome here and I'm banning you from interacting with Homebrew on GitHub.
